---
title: "R Notebook"
output: html_notebook
---
#PhyloSeq
Amplicon bioinformatics : from raw reads to tables

#Loading the data
```{r}
ps_connect <-url("https://raw.githubusercontent.com/spholmes/F1000_workflow/master/data/ps.rds")
ps = readRDS(ps_connect)
ps
```

#Taxonomic filtering
```{r}
# Show available ranks in the dataset
rank_names(ps)
```
We can see the different rank names.

#Create table, number of features for each phyla
```{r}
table(tax_table(ps)[, "Phylum"], exclude = NULL)
```

"NA" should be remove

```{r}
ps <- subset_taxa(ps, !is.na(Phylum) & !Phylum %in% c("", "uncharacterized"))
```

```{r}
# Compute prevalence of each feature, store as data.frame
prevdf = apply(X = otu_table(ps),
               MARGIN = ifelse(taxa_are_rows(ps), yes = 1, no = 2),
               FUN = function(x){sum(x > 0)})
# Add taxonomy and total read counts to this data.frame
prevdf = data.frame(Prevalence = prevdf,
                    TotalAbundance = taxa_sums(ps),
                    tax_table(ps))
```

```{r}
plyr::ddply(prevdf, "Phylum", function(df1){cbind(mean(df1$Prevalence),sum(df1$Prevalence))})
```
We can see the phyla prevalence.

#Filtering
```{r}
# Define phyla to filter
filterPhyla = c("Fusobacteria", "Deinococcus-Thermus")
# Filter entries with unidentified Phylum.
ps1 = subset_taxa(ps, !Phylum %in% filterPhyla)
ps1
```

#Prevalence Filtering
##Unsupervised. This filtering step can be applied even in settings where taxonomic annotation is unavailable or unreliable.
```{r}
# Subset to the remaining phyla
prevdf1 = subset(prevdf, Phylum %in% get_taxa_unique(ps1, "Phylum"))
ggplot(prevdf1, aes(TotalAbundance, Prevalence / nsamples(ps),color=Phylum)) +
  # Include a guess for parameter
  geom_hline(yintercept = 0.05, alpha = 0.5, linetype = 2) +  geom_point(size = 2, alpha = 0.7) +
  scale_x_log10() +  xlab("Total Abundance") + ylab("Prevalence [Frac. Samples]") +
  facet_wrap(~Phylum) + theme(legend.position="none")
```
We explore the relationship of prevalence and total read count for each feature

Figure : Taxa prevalence versus total counts
No natural separation is immediately evident. It could be reasonably to define a prevalence threshold in a range of zero to ten percent or so. 

##Define prevalence threshold ad 5% of total samples
```{r}
prevalenceThreshold = 0.05 * nsamples(ps)
prevalenceThreshold
```
```{r}
keepTaxa = rownames(prevdf1)[(prevdf1$Prevalence >= prevalenceThreshold)]
ps2 = prune_taxa(keepTaxa, ps)
```



#Agglomerate taxa
Useful to agglomerate the data features corresponding to closely related data when there is known to be a lot of species or sub-species functional redundancy in a community.
Taxonomic agglomeration has the advantage of being much easier to define ahead of time. 

##Combine all features that descend from the same genus
How many genera would be present after filtering ?
```{r}
length(get_taxa_unique(ps2, taxonomic.rank = "Genus"))
```
There are 49 genera who would be present after filtering.

```{r}
ps3 = tax_glom(ps2, "Genus", NArm = TRUE)
```


A tree height corresponding to the phylogenetic distance between features that should define their grouping.
```{r}
h1 = 0.4
ps4 = tip_glom(ps2, h = h1)
```

```{r}
multiPlotTitleTextSize = 15
p2tree = plot_tree(ps2, method = "treeonly",
                   ladderize = "left",
                   title = "Before Agglomeration") +
  theme(plot.title = element_text(size = multiPlotTitleTextSize))
p3tree = plot_tree(ps3, method = "treeonly",
                   ladderize = "left", title = "By Genus") +
  theme(plot.title = element_text(size = multiPlotTitleTextSize))
p4tree = plot_tree(ps4, method = "treeonly",
                   ladderize = "left", title = "By Height") +
  theme(plot.title = element_text(size = multiPlotTitleTextSize))
```

```{r}
grid.arrange(nrow = 1, p2tree, p3tree, p4tree)
```
```{r}
install.packages("gridExtra")
```

```{r}
grid.arrange(nrow = 1, p2tree, p3tree, p4tree)
```



